src/processor/common.ts

```diff
@@ -0,0 +1,49 @@
+import { pullAt } from 'lodash';
+import {
+  ExportableAssignmentRule,
+  ExportableCaretValueRule,
+  ExportableSdRule
+} from '../exportable';
+
+// Places general Quantity-setting rules ahead of Quantity.unit setting rules
+export function switchQuantityRules(rules: ExportableSdRule[]): void {
+  const seenRules = new Map();
+  rules.forEach((rule, index) => {
+    if (!(rule instanceof ExportableAssignmentRule || rule instanceof ExportableCaretValueRule)) {
+      return;
+    }
+    if (!seenRules.has(rule.path)) seenRules.set(rule.path, index);
+    const unitRulePath = rule.path.concat('.unit');
+    if (seenRules.has(unitRulePath)) {
+      const unitRuleIndex = seenRules.get(unitRulePath);
+      rules.splice(unitRuleIndex, 0, rule);
+      pullAt(rules, [index + 1]);
+    }
+  });
+}
+
+// Places general Quantity-setting rules ahead of Quantity.unit setting rules
+// Note: A different function is required for Instances due to the way assignment rules are generated
+// When STU2 Features are enabled in GoFSH, this can be removed in favor of editing the
+// CombineCodingAndQuantityValuesOptimizer to include unit rules as well
+export function switchQuantityRulesOnInstances(rules: ExportableSdRule[]): void {
+  let unitIndex: number;
+  let lastSiblingIndex: number;
+  let basePath: string;
+  let siblingPaths: string[];
+  rules.forEach((rule, index) => {
+    if (rule.path.endsWith('.unit')) {
+      unitIndex = index;
+      basePath = rule.path.replace(/\.unit$/, '');
+      siblingPaths = [`${basePath}.system`, `${basePath}.code`, `${basePath}.value`];
```

Comment:
So I think on an Instance, the rules would always come as individual rules, like you are accounting for here. But, I think that the rules could also come as individual rules on a StructureDefinition (if `patternQuantity` is not used, and instead the constraints are individually applied to each element). That makes me think that the processing required for a StructureDefinition is basically a superset of the processing required for an Instance, so I think we could just have one single function, that handles both cases, individual and combined quantity rules. When applied to an instance, the parts of the code handling combined rules would never be hit, but that is fine I think. Does this make sense?

Reply:
This makes perfect sense, I'll combine these into one function. Thank you for thinking of this!